### PR TITLE
Fixed #27775 -- Respect custom expiry in signed cookies.

### PR DESCRIPTION
**Pull request by pelme**
The custom expiry value `_session_expiry` is stored in the session
itself. The signed cookie needs to first be loaded, and then the value
needs to be checked for validity.

This PR calls `signing.loads()` twice which may incur a small performance
penalty. The proper solution would be to refactor the session expiry
as discussed in #19201. However, this fix should not really make that
refactor any harder an may be an acceptable fix in the meantime.

Refs #19200, #19201.

### CODE REVIEW

File: django/contrib/sessions/backends/signed_cookies.py [hunk 1]
```diff
@@ -12,17 +15,34 @@ def load(self):
...
+
             return signing.loads(
                 self.session_key,
                 serializer=self.serializer,
-                # This doesn't handle non-default expiry dates, see #19201
-                max_age=settings.SESSION_COOKIE_AGE,
+                max_age=max_age,
                 salt='django.contrib.sessions.backends.signed_cookies',
             )
+
```
Comment (timgraham) [inline django/contrib/sessions/backends/signed_cookies.py]: Please don't make unrelated whitespace changes.

**Comment by pelme** [comment]
@timgraham now there is a proper ticket for this and I've described the motivation for this patched in the ticket:

https://code.djangoproject.com/ticket/27775

File: django/contrib/sessions/backends/signed_cookies.py [hunk 2]
```diff
@@ -12,11 +15,26 @@ def load(self):
...
+                max_age=None,
+                salt='django.contrib.sessions.backends.signed_cookies',
+            )
+
+            max_age = session_data_non_checked_age.get('_session_expiry')
+            if isinstance(max_age, datetime.datetime):
+                max_age = max_age - timezone.now()
+
+            # Handle None and 0
+            if not max_age:
```
Comment (timgraham) [inline django/contrib/sessions/backends/signed_cookies.py]: Test coverage seems lacking. For example, commenting this if-block out doesn't result in any failures. Could you add some tests as needed for all the changes?

**Comment by virtuallight** [comment]
@pelme I could add some tests to it if you like?

**Comment by timgraham** [comment]
@virtuallight, feel free to add tests and send a new PR. Closing this one due to inactivity.
